### DISCUSSION TITLE
Optimizing factory pattern

### DISCUSSION THREAD

**Discussion by 0xdcota** [Fujicracy/fuji-v2#332] [Ideas]
Last week (Feb 13, 2023) when I was working on defining the vaults that we need to deploy for each collateral-debt asset pair considering different providers, I realized that the permutation or numbers of vaults that we need to deploy is large. 
This means a lot of repeated bytecode that will make us burn an unnecessary and absurd amount of eth, at least when we get to the point of deploying on mainnet. 

This made me think on a way to address this. 

1. We have the vault factories deploy a "master" implementation and all further vaults are simple proxies that keep state but reference the master logic contract from the vault factory. This is similar to upgradeable contract pattern; however it is managed instead from the specific vault factory.
 
I believe the refactor is not as straight forward and will take some time, but perhaps is the true way to scale. @brozorec @pedrovalido 

I include [here](https://soliditydeveloper.com/clonefactory) a refresher on the factory clone pattern.

**Comment by pedrovalido** [comment]
Great suggestion! Completely agree on this. Even if we want to add more vaults in the future, it will be more efficient! 
About the time to implement this, I would suggest we check the roadmap and launch date to get everything on time.

**Comment by brozorec** [comment]
I also agree. We can use OZ's clone library https://docs.openzeppelin.com/contracts/4.x/api/proxy#minimal_clones

We can plan to create a new factory contract "BorrowingVaultFactory2" and use this pattern.
